syfertext/pipeline/matcher.py

```diff
@@ -0,0 +1,452 @@
+
+            processed_patterns = [[("lower", "hello"), ("is_punct", True), ("lower", "world")],
+                                    [("lower", "hello"), ("lower", _RegexPredicate(attr="lower", value="^[Ii](ndia)")]
+                                ]
+        """
+
+        processed_patterns = list()
+
+        for pattern in patterns:
+
+            assert isinstance(pattern, list)
+            cur_pattern = list()
+
+            for token_spec in pattern:
+
+                # Assert it's a dict containing a match pattern
+                # for exactly one token
+                assert isinstance(token_spec, dict)
+                assert len(token_spec) == 1
+
+                for attr, value in token_spec.items():
+
+                    attr = attr.lower()
+
+                    if isinstance(value, dict):
+                        # It is a predicate
+                        predicate = _get_predicates(attr, value)
+                        cur_pattern.append((attr, predicate))
+                    else:
+                        cur_pattern.append((attr, value))
+
+            processed_patterns.append(cur_pattern)
+
+        return processed_patterns
+
+    @staticmethod
+    def simplify(worker: BaseWorker, matcher: "Matcher"):
+        """Simplifies a Matcher object.
+
+        Args:
+            worker (BaseWorker): The worker on which the simplify operation is carried out.
+            matcher (Matcher): the Matcher object to simplify.
+
+        Returns:
+            (tuple): The simplified Matcher object.
+        """
+
+        # Simplify the object properties
+        # TODO: HOW TO SIMPLIFY VOCAB OBJECT ?
+        # IT REQUIRES ACCESS TO THE LANGUAGE OBJECT"S VOCAB
+        vocab = serde._simplify(worker, matcher.vocab)
+        patterns = serde._simplify(worker, matcher._patterns)
+        callbacks = serde._simplify(worker, matcher._callbacks)
+
+        return (vocab, patterns, callbacks)
+
+    @staticmethod
+    def detail(worker: BaseWorker, simple_obj: tuple):
+        """Takes a simplified Matcher object, details it
+           and returns a Matcher object.
+
+        Args:
+            worker (BaseWorker): The worker on which the detail operation is carried out.
+            simple_obj (tuple): The simplified SubPipeline object.
+
+        Returns:
+            (Matcher): The Matcher object.
+        """
+
+        # Unpack the simplified object
+        vocab, patterns, callbacks = simple_obj
+
+        # Detail each property
+        vocab = serde._detail(worker, vocab)
+        patterns = serde._detail(worker, patterns)
+        callbacks = serde._detail(worker, callbacks)
+
+        # Instantiate a Matcher object
+        matcher = Matcher(vocab=vocab)
+
+        # Add patterns to matcher
+        for key, patterns_ in patterns.items():
+            matcher.add(key, patterns_, callbacks.get(key, None))
+
+        return matcher
+
+
+# End of Matcher class
+
+
+def _check_match(token_spec, token):
+    """
```

Comment:
Please add an explanation of what this function is supposed to do. 

(Why don't you put it inside the Matcher class?)